server/src/main/java/org/elasticsearch/index/analysis/IndexAnalyzers.java

```diff
@@ -27,83 +27,142 @@
     }
 
     /**
      * Returns the default search analyzer for this index. If not set, this will return the default analyzer
      */
-    public NamedAnalyzer getDefaultSearchAnalyzer() {
-        return analyzers.getOrDefault(DEFAULT_SEARCH_ANALYZER_NAME, getDefaultIndexAnalyzer());
+    default NamedAnalyzer getDefaultSearchAnalyzer() {
+        NamedAnalyzer analyzer = getAnalyzer(AnalyzerType.ANALYZER, DEFAULT_SEARCH_ANALYZER_NAME);
+        if (analyzer != null) {
+            return analyzer;
+        }
+        return getDefaultIndexAnalyzer();
     }
 
     /**
-     * Returns the default search quote analyzer for this index
+     * Returns the default search quote analyzer for this index. If not set, this will return the default
+     * search analyzer
      */
-    public NamedAnalyzer getDefaultSearchQuoteAnalyzer() {
-        return analyzers.getOrDefault(DEFAULT_SEARCH_QUOTED_ANALYZER_NAME, getDefaultSearchAnalyzer());
+    default NamedAnalyzer getDefaultSearchQuoteAnalyzer() {
+        NamedAnalyzer analyzer = getAnalyzer(AnalyzerType.ANALYZER, AnalysisRegistry.DEFAULT_SEARCH_QUOTED_ANALYZER_NAME);
+        if (analyzer != null) {
+            return analyzer;
+        }
+        return getDefaultSearchAnalyzer();
+    }
+
+    /**
+     * Reload any analyzers that have reloadable components
+     */
+    default List<String> reload(AnalysisRegistry analysisRegistry, IndexSettings indexSettings) throws IOException {
+        return List.of();
+    }
+
+    default void close() throws IOException {}
+
+    static IndexAnalyzers of(Map<String, NamedAnalyzer> analyzers) {
+        return of(analyzers, Map.of(), Map.of(), false);
     }
 
-    @Override
-    public void close() throws IOException {
-        IOUtils.close(
-            Stream.of(analyzers.values().stream(), normalizers.values().stream(), whitespaceNormalizers.values().stream())
-                .flatMap(s -> s)
-                .filter(a -> a.scope() == AnalyzerScope.INDEX)
-                .toList()
-        );
+    static IndexAnalyzers of(Map<String, NamedAnalyzer> analyzers, Map<String, NamedAnalyzer> tokenizers) {
+        return of(analyzers, tokenizers, Map.of(), false);
     }
+
+    static IndexAnalyzers of(
+        Map<String, NamedAnalyzer> analyzers,
+        Map<String, NamedAnalyzer> normalizers,
+        Map<String, NamedAnalyzer> whitespaceNormalizers,
+        boolean closeable
+    ) {
+        return new IndexAnalyzers() {
+            @Override
+            public NamedAnalyzer getAnalyzer(AnalyzerType type, String name) {
+                return switch (type) {
+                    case ANALYZER -> analyzers.get(name);
+                    case NORMALIZER -> normalizers.get(name);
+                    case WHITESPACE -> whitespaceNormalizers.get(name);
+                };
+            }
+
+            @Override
+            public void close() throws IOException {
+                if (closeable) {
```

Comment:
Could you explain quickly (and maybe add javadocs to the interface) when this flag should be set and when not? I see it set to true in MapperServiceFactory but to "false" in the several "IndexAnalyzer.of" methods. Some hin on how to use it in the docs would be great.

Reply:
This is actually a leftover of a previous go at this, I'll remove it.